### PR TITLE
MHC check to respect maxUnhealthy value

### PR DESCRIPTION
**Pull request by Danil-Grigorev**
OCPCLOUD-874 - Implement a validation for MHC maxUnhealthy

### CODE REVIEW

File: pkg/framework/machines.go [hunk 1]
```diff
@@ -103,28 +125,34 @@ func GetMachineFromNode(client runtimeclient.Client, node *corev1.Node) (*mapiv1
...
+			}
 		}
 		return true, nil
 	})
 }
 
-// WaitForMachineDelete polls until the given Machines are not found.
-func WaitForMachineDelete(c client.Client, machines ...*mapiv1beta1.Machine) {
+// WaitForMachinesDeleted polls until the given Machines are not found.
+func WaitForMachinesDeleted(c client.Client, machines ...*mapiv1beta1.Machine) {
```
Comment (Danil-Grigorev) [inline pkg/framework/machines.go:142]: Waits for https://github.com/openshift/cluster-api-actuator-pkg/pull/159 functionality to be merged.

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 2]
```diff
@@ -101,24 +88,95 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+			Name:         msName,
+			Replicas:     3,
+			ProviderSpec: providerSpec,
+			Labels: map[string]string{
+				"mhc.framework.openshift.io": msName,
+				framework.ClusterKey:         clusterName,
+			},
+		}
+
+		By("Creating a new MachineSet")
```
Comment (alexander-demichev) [inline pkg/machinehealthcheck/machinehealthcheck.go:126]: Creating and deleting machineset is done in `BeforeEach()` and `AfterEach()`

Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go]: https://deck-ci.apps.ci.l2s4.p1.openshiftapps.com/view/gcs/origin-ci-test/pr-logs/pull/openshift_cluster-api-actuator-pkg/160/pull-ci-openshift-cluster-api-actuator-pkg-master-e2e-gcp-operator/195 What else could cause this, but a double create call in a parallel execution of `BeforeEach` with a create call. There is definitely a race condition somewhere. Something like `BeforeAll`/`AfterAll` would be perfect

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 3]
```diff
@@ -46,10 +46,6 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 		providerSpec := workers[0].Spec.Template.Spec.ProviderSpec.DeepCopy()
 		clusterName := workers[0].Spec.Template.Labels[framework.ClusterKey]
 
-		// TODO(bison): This should probably be appended with
-		// something other than a timestamp, e.g. a random string.
-		msName := fmt.Sprintf("e2e-mhc-%d", time.Now().Unix())
```
Comment (elmiko) [inline pkg/machinehealthcheck/machinehealthcheck.go:51]: i see that msName has moved to the outer scope, but i am wondering about this TODO note. do we need to have some sort random string on the name?

File: pkg/infra/infra.go [hunk 4]
```diff
@@ -229,12 +229,10 @@ var _ = Describe("[Feature:Machines] Managed cluster should", func() {
 		Expect(err).ToNot(HaveOccurred())
 		Expect(machines).ToNot(BeEmpty())
 
-		machine := machines[0]
-
-		By(fmt.Sprintf("deleting machine object %q", machine.Name))
-		err = framework.DeleteMachine(client, machine)
+		By(fmt.Sprint("deleting all machines"))
+		err = framework.DeleteMachines(client, machines...)
```
Comment (enxebre) [inline pkg/infra/infra.go:233]: how is this change related to this PR?

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 5]
```diff
@@ -118,7 +115,51 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+		Expect(machines).ToNot(BeEmpty())
+
+		for _, machine := range machines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			err = framework.AddNodeCondition(client, node, nodeCondition)
+			Expect(err).ToNot(HaveOccurred())
+		}
+
+		By("Creating a MachineHealthCheck resource with low maxUnhealthy threshold")
```
Comment (enxebre) [inline pkg/machinehealthcheck/machinehealthcheck.go:133]: what is low? can we be explicit?

Comment (enxebre) [reply pkg/machinehealthcheck/machinehealthcheck.go:133]: can we do something more disruptive with a higher value for maxUnhealthy? wait for a few machines to be remediated and see the MHC short circuit for the others?

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 6]
```diff
@@ -20,7 +18,9 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 
 	var machineSet *mapiv1beta1.MachineSet
 	var machineSetParams framework.MachineSetParams
+	var maxAllowedUnhealthy = 1
```
Comment (enxebre) [inline pkg/machinehealthcheck/machinehealthcheck.go:21]: can we name this consistently with api field, i.e `MaxUnhealthy`

Comment (enxebre) [reply pkg/machinehealthcheck/machinehealthcheck.go:21]: also should it be rather a constant?

File: pkg/framework/machines.go [hunk 7]
```diff
@@ -44,6 +44,28 @@ func GetMachine(c client.Client, name string) (*mapiv1beta1.Machine, error) {
 	return machine, nil
 }
 
+// MachinesPresent looks for matching machine in a predefined list
+func MachinesPresent(existingMachines []*mapiv1beta1.Machine, machines ...*mapiv1beta1.Machine) bool {
+	if len(existingMachines) < len(machines) {
+		return false
+	}
+	for _, machine := range machines {
```
Comment (enxebre) [inline pkg/framework/machines.go:52]: this nested loops would be cheaper in terms of complexity if you have two independent loops over a map and hash objects by UID.

File: pkg/framework/machines.go [hunk 8]
```diff
@@ -44,6 +44,28 @@ func GetMachine(c client.Client, name string) (*mapiv1beta1.Machine, error) {
 	return machine, nil
 }
 
+// MachinesPresent looks for matching machine in a predefined list
```
[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/framework/machines.go:47]: You are right, the comment message is completely misleading. It is supposed to check that each machine in `machines` list is present in provided `existingMachines` list.

File: pkg/infra/infra.go (hunk 4 above)
Comment (Danil-Grigorev) [reply pkg/infra/infra.go:233]: It is the same thing, as in comment https://github.com/openshift/cluster-api-actuator-pkg/pull/160/files#r431705540, will completely exclude this and rebase after merging. Letft it here in hope to see CI jobs passing.

File: pkg/machinehealthcheck/machinehealthcheck.go (hunk 3 above)
Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go:51]: I decided to move onto `GenerateName` which will have the same effect here, and will give a unique name for the machine - https://github.com/openshift/cluster-api-actuator-pkg/pull/160/files#diff-c7d1003b84db4b5057508a81ab203c46R80

File: pkg/machinehealthcheck/machinehealthcheck.go (hunk 6 above)
Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go]: It should, but I can't pass a pointer to a constant.

File: pkg/machinehealthcheck/machinehealthcheck.go (hunk 5 above)
Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go]: I extended the case and made it more visible, so it will ensures the machines will be remediated when the number of unhealthy entities does not reach the threshold, and won't be past that. Also added checks for all events being emitted during that procedure.

File: pkg/framework/framework.go [hunk 9]
```diff
@@ -176,3 +177,26 @@ func WaitForValidatingWebhook(client runtimeclient.Client, name string) bool {
...
 	return true
 }
+
+// WaitForEvent expects to find the given event in a reasonable timeframe
+func WaitForEvent(c runtimeclient.Client, kind, name, reason string) error {
+	return wait.PollImmediate(RetryMedium, WaitShort, func() (bool, error) {
+		eventList := corev1.EventList{}
+		if err := c.List(context.TODO(), &eventList); err != nil {
+			glog.Errorf("error querying api for eventList object: %v, retrying...", err)
+			return false, nil
```
[the remark this answers is not in the document]
Comment (elmiko) [reply pkg/framework/framework.go:187]: @Danil-Grigorev you marked this as resolved, but i don't see any change or discussion. what's the story?

[the remark this answers is not in the document]
Comment (Danil-Grigorev) [reply pkg/framework/framework.go:187]: Sorry, I picked the change, but returning the error here was causing CI to fail, so I reverted it. Especially in a long running task like this, returning an error which aborts the loop is dangerous...

[the remark this answers is not in the document]
Comment (elmiko) [reply pkg/framework/framework.go:187]: weird that it was causing failure since you still return false as well. but, no worries, thanks for the update!

File: pkg/framework/framework.go [hunk 10]
```diff
@@ -176,3 +177,26 @@ func WaitForValidatingWebhook(client runtimeclient.Client, name string) bool {
 
 	return true
 }
+
+// WaitForEvent expects to find the given event
+func WaitForEvent(c runtimeclient.Client, kind, name, reason string) error {
+	return wait.PollImmediate(RetryMedium, WaitMedium, func() (bool, error) {
+		eventList := corev1.EventList{}
+		if err := c.List(context.TODO(), &eventList); err != nil {
```
Comment (JoelSpeed) [inline pkg/framework/framework.go:185]: Use background for tests where we don't think we will replace it at a later date
```suggestion
		if err := c.List(context.Background(), &eventList); err != nil {
```

File: pkg/framework/machines.go [hunk 11]
```diff
@@ -44,6 +45,24 @@ func GetMachine(c client.Client, name string) (*mapiv1beta1.Machine, error) {
...
+// MachinesPresent search for each provided machine in `machines` argument in the predefined `existingMachines` list
+// and returns true when all of them were found
+func MachinesPresent(existingMachines []*mapiv1beta1.Machine, machines ...*mapiv1beta1.Machine) bool {
+	if len(existingMachines) < len(machines) {
+		return false
+	}
+	existingMachinesMap := map[types.UID]bool{}
+	for _, existing := range existingMachines {
+		existingMachinesMap[existing.UID] = true
+	}
```
Comment (JoelSpeed) [inline pkg/framework/machines.go:57]: Technically this is more efficient, but what you have is fine, just as an education point, change if you want. A bool takes up more memory than an empty struct
```suggestion
	existingMachinesMap := map[types.UID]struct{}{}
	for _, existing := range existingMachines {
		existingMachinesMap[existing.UID] = struct{}
	}
```

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 12]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
 	})
 
 	AfterEach(func() {
+		By("Deleting the MachineHealthCheck resource")
+		err := client.Delete(context.Background(), machinehealthcheck)
+		Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:54]: I would have inlined this, but as it matches the existing, can leave it (or update both) `Expect(client.Delete(...)).To(Succeed()`

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 13]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
 
-		By("Verifying the matching Machine is deleted")
-		framework.WaitForMachineDelete(client, machine)
+		By("Waiting for each unhealthy machine to be deleted")
+		framework.WaitForMachinesDeleted(client, unhealthyMachines...)
+
+		By("Waiting for MachineDeleted event from MachineHealthCheck for each unhealthy machine")
+		for _, machine := range unhealthyMachines {
+			err = framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")
+			Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:102]: Could inline this? That way we don't shadow the err variable
```suggestion
			Expect(framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")).To(Succeed())
```

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 14]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machines).ToNot(BeEmpty())
+
+		By("Setting unhealthy conditions on machine nodes, but exceding maxUnhealthy threshold")
+		unhealthyMachines := machines[:maxUnhealthy+1]
+		for _, machine := range unhealthyMachines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			err = framework.AddNodeCondition(client, node, nodeCondition)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:121]: Could inline this

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 15]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+			},
+			MaxUnhealthy: &maxUnhealthy,
+		}
+
+		machinehealthcheck, err = framework.CreateMHC(client, mhcParams)
+		Expect(err).ToNot(HaveOccurred())
+		Expect(machinehealthcheck).ToNot(BeNil())
+
+		By("Waiting for RemediationRestricted event from MachineHealthCheck")
+		err = framework.WaitForEvent(client, "MachineHealthCheck", machineSet.Name, "RemediationRestricted")
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:143]: Could inline this

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 16]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
 		Expect(err).ToNot(HaveOccurred())
-	})
 
-	// It("should respect maxUnhealthy", func() {
-	// 	// TODO(bison): This.
-	// })
+		By("Ensuring none of the unhealthy machines are deleted")
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:146]: Should this be healthy rather than unhealthy?

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 17]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+		untouchedMachines, err := framework.GetMachines(client, &selector)
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:147]: Perhaps healthy rather than untouched, untouched seems a bit of a weird description to me 😅

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 18]
```diff
@@ -68,57 +49,105 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
-
-		err = framework.AddNodeCondition(client, node, nodeCondition)
-		Expect(err).ToNot(HaveOccurred())
+		By("Setting unhealthy conditions on machine nodes, but not exceding maxUnhealthy threshold")
+		unhealthyMachines, healthyMachines := machines[:maxUnhealthy-1], machines[maxUnhealthy-1:]
+		for _, machine := range unhealthyMachines {
+			node, err := framework.GetNodeForMachine(client, machine)
+			Expect(err).ToNot(HaveOccurred())
+			err = framework.AddNodeCondition(client, node, nodeCondition)
+			Expect(err).ToNot(HaveOccurred())
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:73]: Potentially inline this

File: pkg/machinehealthcheck/machinehealthcheck.go (hunk 2 above)
Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go]: Realized that it is no longer needed.

File: pkg/framework/machinesets.go [hunk 19]
```diff
@@ -47,39 +47,31 @@ func BuildMachineSetParams(client runtimeclient.Client, replicas int) MachineSet
...
 // CreateMachineSet creates a new MachineSet resource.
 func CreateMachineSet(c client.Client, params MachineSetParams) (*mapiv1beta1.MachineSet, error) {
-	if params.Labels == nil {
-		params.Labels = make(map[string]string)
-	}
-
-	// TODO(bison): It would be nice to automatically set the Cluster ID / name
-	// in the labels here, but I'm not sure how to easily find it.
-	params.Labels["e2e.openshift.io"] =
-		fmt.Sprintf("%s-%d", params.Name, time.Now().Unix())
```
Comment (JoelSpeed) [inline pkg/framework/machinesets.go:72]: Will this effect other tests? How many of the tests use `CreateMachineSet`?

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 20]
```diff
@@ -68,57 +49,104 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
-		Expect(mhc).ToNot(BeNil())
+		Expect(machinehealthcheck).ToNot(BeNil())
+
+		By("Waiting for each unhealthy machine to be deleted")
+		framework.WaitForMachinesDeleted(client, unhealthyMachines...)
+
+		By("Waiting for MachineDeleted event from MachineHealthCheck for each unhealthy machine")
+		for _, machine := range unhealthyMachines {
+			Expect(framework.WaitForEvent(client, "Machine", machine.Name, "MachineDeleted")).ToNot(HaveOccurred())
+		}
```
Comment (JoelSpeed) [inline pkg/machinehealthcheck/machinehealthcheck.go:99]: Should these be the other way around, since the events will happen before the machines are deleted right?

Comment (Danil-Grigorev) [reply pkg/machinehealthcheck/machinehealthcheck.go:99]: This happens roughly at the same time, and the events are persisted for some time, so I left it there. Also, if the machines are not deleted, we'd rather see the timeout on machine deletion, then on some event search in my opinion. It is better for debugging imo

File: pkg/framework/machinesets.go (hunk 19 above)
Comment (Danil-Grigorev) [reply pkg/framework/machinesets.go:72]: Yes, it will ensure we will never see a race condition, when the CreateMachineSet was called at the same time, so the labels on two machines would be matching, and there will be no good way for tests to find out how many machines the machine et actually spawned

Comment (JoelSpeed) [reply pkg/framework/machinesets.go:72]: Sounds like a good improvement 👍

File: pkg/machinehealthcheck/machinehealthcheck.go (hunk 20 above)
Comment (JoelSpeed) [reply pkg/machinehealthcheck/machinehealthcheck.go:99]: That's a reasonable argument, sounds good to me :)

File: pkg/machinehealthcheck/machinehealthcheck.go [hunk 21]
```diff
@@ -68,57 +49,104 @@ var _ = Describe("[Feature:MachineHealthCheck] MachineHealthCheck", func() {
...
+		Expect(allMachines).ToNot(BeEmpty())
+		Expect(framework.MachinesPresent(allMachines, healthyMachines...)).To(BeTrue())
 
 		By("Verifying the MachineSet recovers")
 		framework.WaitForMachineSet(client, machineSet.GetName())
+	})
 
-		By("Deleting the MachineHealthCheck resource")
-		err = client.Delete(context.Background(), mhc)
+	It("should not remediate larger number of unhealthy machines then maxUnhealthy", func() {
```
Comment (elmiko) [inline pkg/machinehealthcheck/machinehealthcheck.go:111]: this should be "than" rather then "then", but it's not a blocker for me.
